pgpu-render/src/lib.rs

```diff
@@ -98,24 +101,239 @@ pub unsafe extern "C" fn pgpu_scene_destroy(scene: *mut PgpuScene) {
+    pub yx: f32,
+    pub xy: f32,
+    pub yy: f32,
+    pub dx: f32,
+    pub dy: f32,
+}
+
+impl From<PgpuTransform> for PgpuAffine {
+    fn from(xform: PgpuTransform) -> Self {
+        Self {
+            xx: xform.xx,
+            yx: xform.yx,
+            xy: xform.xy,
+            yy: xform.yy,
+            dx: xform.dx,
+            dy: xform.dy,
+        }
+    }
+}
+
+pub type PgpuAffine = piet_scene::geometry::Affine;
+
 /// Creates a new builder for filling a piet-gpu scene. The specified scene
 /// should not be accessed while the builder is live.
 #[no_mangle]
-pub unsafe extern "C" fn pgpu_scene_builder_new(
+pub unsafe extern "C" fn pgpu_scene_builder_for_scene(
     scene: *mut PgpuScene,
 ) -> *mut PgpuSceneBuilder<'static> {
-    Box::into_raw(Box::new((*scene).build()))
+    Box::into_raw(Box::new((*scene).builder()))
+}
+
+/// Creates a new builder for filling a piet-gpu scene fragment. The specified
+/// scene fragment should not be accessed while the builder is live.
+#[no_mangle]
+pub unsafe extern "C" fn pgpu_scene_builder_for_fragment(
+    fragment: *mut PgpuSceneFragment,
+) -> *mut PgpuSceneBuilder<'static> {
+    Box::into_raw(Box::new((*fragment).builder()))
 }
 
 /// Adds a glyph with the specified transform to the underlying scene.
 #[no_mangle]
 pub unsafe extern "C" fn pgpu_scene_builder_add_glyph(
     builder: *mut PgpuSceneBuilder<'static>,
     glyph: *const PgpuGlyph,
-    transform: &[f32; 6],
+    transform: *const PgpuTransform,
 ) {
-    let transform = piet_scene::geometry::Affine::new(transform);
-    (*builder).add_glyph(&*glyph, &transform);
+    (*builder).add_glyph(&*glyph, &(*transform).into());
+}
+
+impl Iterator for PgpuPathIter {
+    type Item = piet_scene::path::Element;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut el = PgpuPathElement {
+            verb: PgpuPathVerb::MoveTo,
+            points: [PgpuPoint::default(); 3],
+        };
+        if (self.next_element)(self.context, &mut el as _) {
+            let p = &el.points;
+            Some(match el.verb {
+                PgpuPathVerb::MoveTo => Element::MoveTo((p[0].x, p[0].y).into()),
+                PgpuPathVerb::LineTo => Element::LineTo((p[0].x, p[0].y).into()),
+                PgpuPathVerb::QuadTo => {
+                    Element::QuadTo((p[0].x, p[0].y).into(), (p[1].x, p[1].y).into())
+                }
+                PgpuPathVerb::CurveTo => Element::CurveTo(
+                    (p[0].x, p[0].y).into(),
+                    (p[1].x, p[1].y).into(),
+                    (p[2].x, p[2].y).into(),
+                ),
+                PgpuPathVerb::Close => Element::Close,
+            })
+        } else {
+            None
+        }
+    }
+}
+
+/// Sets the current absolute transform for the scene builder.
+#[no_mangle]
+pub unsafe extern "C" fn pgpu_scene_builder_transform(
+    builder: *mut PgpuSceneBuilder<'static>,
+    transform: *const PgpuTransform,
+) {
+    if !transform.is_null() {
```

Comment:
I'm wondering if [as_ref](https://doc.rust-lang.org/std/primitive.pointer.html#method.as_ref) here and in a few other places might be more concise than the is_null check.

Reply:
Oh nice. I wasn't aware of that method. I'm happy to clean those up by I'd like to land this as is to avoid merge conflicts with #189. I can either add these changes as part of that PR or, probably better, do an FFI cleanup as a follow up to avoid making that one any larger.